perf/impl/go-libp2p/v0.27/perf.go

```diff
@@ -0,0 +1,171 @@
+package main
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	pool "github.com/libp2p/go-buffer-pool"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	msmux "github.com/multiformats/go-multistream"
+)
+
+var log = logging.Logger("perf")
+
+const (
+	BlockSize = 64 << 10
+
+	ID = "/perf/1.0.0"
+)
+
+type PerfService struct {
+	Host host.Host
+}
+
+func NewPerfService(h host.Host) *PerfService {
+	ps := &PerfService{h}
+	h.SetStreamHandler(ID, ps.PerfHandler)
+	return ps
+}
+
+func (p *PerfService) PerfHandler(s network.Stream) {
+	buf := pool.Get(BlockSize)
+	defer pool.Put(buf)
+
+	u64Buf := make([]byte, 8)
+	_, err := io.ReadFull(s, u64Buf)
+	if err != nil {
+		log.Errorw("err", err)
+		s.Reset()
+		return
+	}
+
+	bytesToSend := binary.BigEndian.Uint64(u64Buf)
+
+	_, err = p.drainStream(context.Background(), s, buf)
+	if err != nil {
+		log.Errorw("err", err)
+		s.Reset()
+		return
+	}
+
+	err = p.sendBytes(context.Background(), s, bytesToSend, buf)
+	if err != nil {
+		log.Errorw("err", err)
+		s.Reset()
+		return
+	}
+
+}
+
+func (ps *PerfService) sendBytes(ctx context.Context, s network.Stream, bytesToSend uint64, buf []byte) error {
+	for bytesToSend > 0 {
+		toSend := buf
+		if bytesToSend < BlockSize {
+			toSend = buf[:bytesToSend]
+		}
+
+		n, err := s.Write(toSend)
+		if err != nil {
+			return err
+		}
+		bytesToSend -= uint64(n)
+	}
+	s.CloseWrite()
+
+	return nil
+}
+
+func (ps *PerfService) drainStream(ctx context.Context, s network.Stream, buf []byte) (uint64, error) {
+	var recvd uint64
+	for {
+		n, err := s.Read(buf)
+		recvd += uint64(n)
+		if err == io.EOF {
+			return recvd, nil
+		} else if err != nil {
+			s.Reset()
+			return recvd, err
+		}
+	}
+}
+
+func (ps *PerfService) RunPerf(ctx context.Context, p peer.ID, bytesToSend uint64, bytesToRecv uint64) (time.Duration, time.Duration, error) {
+	// Use ps.Host.Network().NewStream() instead of ps.Host.NewStream() to
+	// skip waiting for identify protocol to finish.
+	s, err := ps.Host.Network().NewStream(network.WithNoDial(ctx, "already dialed"), p)
```

Comment:
Fine by me. I did the following reverts, thus negotiating both Noise and TLS (no additional roundtrip due to https://github.com/libp2p/go-libp2p/issues/2330 preventing optimistic negotiation) and waiting for the identify exchange on connection establishment and stream creation. Now, it works just like a usual Kubo setup.

- 13e244b perf Revert "fix(go-libp2p): use tls only"
- a8ded97 Revert "fix(go-libp2p): use NewWithoutDefaults to disable Noise"
- 88b6f34 Revert "fix(go-libp2p): don't wait for identify round trip"
- 6e0fc1d Revert "fix(go-libp2p): don't wait for identify when dialing"

//CC @MarcoPolo since we discussed not waiting on identify here.

> I'd argue that we should do the latter, and take this as an opportunity to change the default behavior (which I've wanted to do for a long time). Maybe putting actual numbers on it, and being to claim credit for an improvement in our new metrics page, will make prioritizing this fix more attractive :)

:+1: That makes sense to me @marten-seemann.